### DISCUSSION TITLE
httpcore.RemoteProtocolError: Server disconnected without sending a response

### DISCUSSION THREAD

**Discussion by Pedro-Leitek** [AxisCommunications/discussions#45] [VAPIX] [answered]
Hi,

I'm implementing an application that uses an axis camera. The app needs to work 24/7. The thing is, after 2-3 hours the following error appears and the app crashes. I'm using the python version of the api. Apparently the camera disconnects from the main application.
Can you help me?

Thanks

Exception in thread Thread-1:
Traceback (most recent call last):
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpx/_transports/default.py", line 62, in map_httpcore_exceptions
yield
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpx/_transports/default.py", line 281, in handle_async_request
extensions=request.extensions,
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpcore/_async/connection_pool.py", line 235, in handle_async_request
method, url, headers=headers, stream=stream, extensions=extensions
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpcore/_async/connection.py", line 149, in handle_async_request
method, url, headers, stream, extensions
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpcore/_async/http11.py", line 128, in handle_async_request
) = await self._receive_response(timeout)
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpcore/_async/http11.py", line 189, in _receive_response
event = await self._receive_event(timeout)
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpcore/_async/http11.py", line 236, in _receive_event
raise RemoteProtocolError(msg)
httpcore.RemoteProtocolError: Server disconnected without sending a response.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "/home/ai-dev/AI-Engine/Projeto_6_Videre/VidereApp/axis/vapix.py", line 268, in request
method, url, auth=self.auth, timeout=TIME_OUT, **kwargs # type: ignore [arg-type]
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpx/_client.py", line 1484, in request
request, auth=auth, follow_redirects=follow_redirects
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpx/_client.py", line 1575, in send
history=[],
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpx/_client.py", line 1602, in _send_handling_auth
history=history,
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpx/_client.py", line 1636, in _send_handling_redirects
response = await self._send_single_request(request)
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpx/_client.py", line 1673, in _send_single_request
response = await transport.handle_async_request(request)
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpx/_transports/default.py", line 281, in handle_async_request
extensions=request.extensions,
File "/usr/lib/python3.6/contextlib.py", line 99, in exit
self.gen.throw(type, value, traceback)
File "/home/ai-dev/.local/lib/python3.6/site-packages/httpx/_transports/default.py", line 79, in map_httpcore_exceptions
raise mapped_exc(message) from exc
httpx.RemoteProtocolError: Server disconnected without sending a response.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
self.run()
File "/usr/lib/python3.6/threading.py", line 864, in run
self._target(self._args, **self._kwargs)
File "utilizador3.py", line 565, in stream
go_azi(p,15.0i)
File "utilizador3.py", line 115, in go_azi
p.move_go_azi(p.loop, azimuth)
File "utilizador3.py", line 264, in move_go_azi
loop.run_until_complete(self.aux_go_azi(self.p,azimuth))
File "/usr/lib/python3.6/asyncio/base_events.py", line 484, in run_until_complete
return future.result()
File "utilizador3.py", line 214, in aux_go_azi
await p.control(camera=1, pan=azi)
File "/home/ai-dev/AI-Engine/Projeto_6_Videre/VidereApp/axis/ptz.py", line 236, in control
return await self._request("post", URL, data=data)
File "/home/ai-dev/AI-Engine/Projeto_6_Videre/VidereApp/axis/vapix.py", line 298, in request
raise RequestError("Connection error: {}".format(errc))
axis.errors.RequestError: Connection error: Server disconnected without sending a response.

**Comment by mattias-kindborg-at-work** [comment, marked as answer]
Hi @Pedro-Leitek.

Could you provide us with a minimal application showcasing the issue. Create a repository here on GitHub with the source code and complete instructions on how to reproduce the issue. Thanks!

**Reply by Pedro-Leitek** [reply]
Hi,

How can I create a repository in your github? The code is company proprietary, I can't show it publicaly.

Thanks

**Reply by mattias-kindborg-at-work** [reply]
Hi.

I understand your problem, but in order to help you I need to be able to reproduce your issue.

**Reply by Pedro-Leitek** [reply]
Ok, here is the code. The idea is to put the camera rotate step by step and, in my application, process a frame for each position.

`
class Camara:
    def __init__(self, lnk, c_nome):#filtros
        self.nome = c_nome
        self.lnk = lnk; 
        self.net = cv2.dnn.readNetFromONNX(config.yoloPathWeights)
        self.framecurrente = None
        self.tempoInicial = time.time()  # Tempo inicial da contagem para guardar a proxima frame da BD
        self.tempoPassado = 0
        self.thread = None
        self.url = urlparse(lnk)
        self.retoques = True
        self.brilho = 0
        self.contraste = 0
        self.mode="manual"
        self.tour=[]

    def stream(self,p):
        connected1=False
        global s1
        while not connected1:  
            # attempt to reconnect, otherwise sleep for 2 seconds  
           try:
              print("three")   
              s1.connect((HOST2, PORT2))
              s1.sendall(json.dumps({"id":1}).encode('utf-8'))
              accept1=s1.recv(1024).decode('ascii')
              if accept1=='True':
                 connected1=True
              else:
                 s1.close()
                 s1 = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                 time.sleep( 2 )
           except socket.error:  
              time.sleep( 2 )

        imagem=cv2.VideoCapture(self.lnk, 0)
        init_azi(p)
        time.sleep(0.8)
        init_ele(p)
        time.sleep(0.8)
        init_zoom_in_out(p)
        time.sleep(0.8)
        get_pos(p)   
        while True: 
          if self.mode=="manual":         
            ativo, frame = imagem.read()
            cv2.rectangle(frame, (0,0), (frame.shape[1],60), (0,0,0), -1)
            cv2.putText(frame, "Leitek-Demo1" + " " + "Time:" + " " + time.strftime('%d-%m-%Y %H:%M:%S', time.localtime(time.time())) +  " " + "Azimuth:" +  " " + 					str(p.azimuth) + " " + "Elevation:"+ " " + str(p.elevation), (0, 45), 		 cv2.FONT_HERSHEY_DUPLEX, 1.25, (255, 255, 255), 2, lineType=cv2.LINE_AA)
            self.framecurrente=frame
          else:
            zoom_in_out(p,2250) #2250
            time.sleep(0.8)
            while self.mode=="auto":
               for i in self.tour:
                  if self.mode=="manual":
                     break
                  array=i.split("_")
                  go_azi(p,(int(array[0])-1)*15.0)
                  time.sleep(0.8)
                  go_ele(p,-9.0*(int(array[1])-1))
                  time.sleep(0.8)
                  get_pos(p)
                  self.processa(p)
            imagem=cv2.VideoCapture(self.lnk, 0)

def go_azi(p,azimuth):     
    p.move_go_azi(p.loop, azimuth)

def go_ele(p,elevation):
    p.move_go_ele(p.loop, elevation)

def get_pos(p):
    pos=p.get_position(p.loop)
    x=pos.split("\n")
    pan=x[0].split("pan=")
    tilt=x[1].split("tilt=")
    if float(pan[1])<0.0:
       p.azimuth=360+float(pan[1])
    else:
       p.azimuth=float(pan[1])
    p.elevation=float(tilt[1])


class PTZ:
    def __init__(self,lnk,user):
         self.url = urlparse(lnk)
         self.username=self.url.username
         self.password=self.url.password
         self.host=self.url.hostname
         self.loop = asyncio.new_event_loop()
         self.p=self.create_ptz(self.loop)
         self.user=user
         print(self.username)
         print(self.password)
         print(self.host)
         self.angle=0.0
         self.azimuth=0.0
         self.elevation=0.0
         self.zoom=1
         self.direction=""


    async def device_creation(self):
        device = await self.axis_device(self.host, 80, self.username, self.password)
        p = self.ptz_control(device)
        return p

    def create_ptz(self,loop):
        asyncio.set_event_loop(loop)
        aux=loop.run_until_complete(self.device_creation())
        return aux

    def ptz_control(self,axis_device) -> PtzControl:
        """Returns the PTZ control mock object."""
        return PtzControl(axis_device.vapix.request)

    async def axis_device(self,
    host: str, port: int, username: str, password: str
    ) -> axis.AxisDevice:
        """Create a Axis device."""
        session = AsyncClient(verify=False)
        device = axis.AxisDevice(
            axis.configuration.Configuration(
                session, host, port=port, username=username, password=password
            )
        )
        try:
            with async_timeout.timeout(5):
                await device.vapix.initialize_users()
                await device.vapix.load_user_groups()
            await device.vapix.initialize_event_instances()

            return device

        except axis.Unauthorized:
            LOGGER.warning(
            "Connected to device at %s but not registered or user not admin.", host
        )

        except (asyncio.TimeoutError, axis.RequestError):
            LOGGER.error("Error connecting to the Axis device at %s", host)

        except axis.AxisException:
            LOGGER.exception("Unknown Axis communication error occurred")

        return device

    def move_go_azi(self,loop,azimuth):
        asyncio.set_event_loop(loop)  
        loop.run_until_complete(self.aux_go_azi(self.p,azimuth))

    def move_go_ele(self,loop,elevation):
        asyncio.set_event_loop(loop)  
        loop.run_until_complete(self.aux_go_ele(self.p,elevation))

    def get_position(self,loop):
         asyncio.set_event_loop(loop)  
         return loop.run_until_complete(self.aux_get_position(self.p))

    async def aux_get_position(self,p):
         return await p.query("position")

    async def aux_go_azi(self,p,azimuth):
        if float(azimuth) > 180.0:
           azi=-360.0 + float(azimuth)
        else:
           azi=float(azimuth)
        await p.control(camera=1, pan=azi)

    async def aux_go_ele(self,p,elevation):
        await p.control(camera=1, tilt=float(elevation))
`

**Reply by mattias-kindborg-at-work** [reply]
The formatting seems the be a bit off. Can you please verify that your code looks like this. For help with formatting code on GitHub, please see [GitHub documentation](https://docs.github.com/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks).

**camera.py**

```python
class Camera:
    def init(self, lnk, c_nome):  # filtros
        self.nome = c_nome
        self.lnk = lnk
        self.net = cv2.dnn.readNetFromONNX(config.yoloPathWeights)
        self.framecurrente = None
        # Tempo inicial da contagem para guardar a proxima frame da BD
        self.tempoInicial = time.time()
        self.tempoPassado = 0
        self.thread = None
        self.url = urlparse(lnk)
        self.retoques = True
        self.brilho = 0
        self.contraste = 0
        self.mode = "manual"
        self.tour = []

    def stream(self, p):
        connected1 = False
        global s1
        while not connected1:
            # attempt to reconnect, otherwise sleep for 2 seconds
            try:
                print("three")
                s1.connect((HOST2, PORT2))
                s1.sendall(json.dumps({"id": 1}).encode('utf-8'))
                accept1 = s1.recv(1024).decode('ascii')
                if accept1 == 'True':
                    connected1 = True
                else:
                    s1.close()
                    s1 = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                    time.sleep(2)
            except socket.error:
                time.sleep(2)

        imagem = cv2.VideoCapture(self.lnk, 0)
        init_azi(p)
        time.sleep(0.8)
        init_ele(p)
        time.sleep(0.8)
        init_zoom_in_out(p)
        time.sleep(0.8)
        get_pos(p)
        while True:
            if self.mode == "manual":
                ativo, frame = imagem.read()
                cv2.rectangle(
                    frame, (0, 0), (frame.shape[1], 60), (0, 0, 0), -1)
                cv2.putText(frame, "Leitek-Demo1" + " " + "Time:" + " " + time.strftime('%d-%m-%Y %H:%M:%S', time.localtime(time.time())) + " " + "Azimuth:" + " " + str(
                    p.azimuth) + " " + "Elevation:" + " " + str(p.elevation), (0, 45),    cv2.FONT_HERSHEY_DUPLEX, 1.25, (255, 255, 255), 2, lineType=cv2.LINE_AA)
                self.framecurrente = frame
            else:
                zoom_in_out(p, 2250)  # 2250
                time.sleep(0.8)
                while self.mode == "auto":
                    for i in self.tour:
                        if self.mode == "manual":
                            break
                        array = i.split("_")
                        go_azi(p, (int(array[0])-1)*15.0)
                        time.sleep(0.8)
                        go_ele(p, -9.0*(int(array[1])-1))
                        time.sleep(0.8)
                        get_pos(p)
                        self.processa(p)
                    imagem = cv2.VideoCapture(self.lnk, 0)

    def go_azi(p, azimuth):
        p.move_go_azi(p.loop, azimuth)

    def go_ele(p, elevation):
        p.move_go_ele(p.loop, elevation)

    def get_pos(p):
        pos = p.get_position(p.loop)
        x = pos.split("\n")
        pan = x[0].split("pan=")
        tilt = x[1].split("tilt=")
        if float(pan[1]) < 0.0:
            p.azimuth = 360+float(pan[1])
        else:
            p.azimuth = float(pan[1])
            p.elevation = float(tilt[1])
```

**ptz.py**

```python
class PTZ:
    def init(self, lnk, user):
        self.url = urlparse(lnk)
        self.username = self.url.username
        self.password = self.url.password
        self.host = self.url.hostname
        self.loop = asyncio.new_event_loop()
        self.p = self.create_ptz(self.loop)
        self.user = user
        print(self.username)
        print(self.password)
        print(self.host)
        self.angle = 0.0
        self.azimuth = 0.0
        self.elevation = 0.0
        self.zoom = 1
        self.direction = ""

    async def device_creation(self):
        device = await self.axis_device(self.host, 80, self.username, self.password)
        p = self.ptz_control(device)
        return p

    def create_ptz(self, loop):
        asyncio.set_event_loop(loop)
        aux = loop.run_until_complete(self.device_creation())
        return aux

    def ptz_control(self, axis_device) -> PtzControl:
        """Returns the PTZ control mock object."""
        return PtzControl(axis_device.vapix.request)

    async def axis_device(self, host: str, port: int, username: str, password: str) -> axis.AxisDevice:
        """Create a Axis device."""
        session = AsyncClient(verify=False)
        device = axis.AxisDevice(
            axis.configuration.Configuration(
                session, host, port=port, username=username, password=password
            )
        )
        try:
            with async_timeout.timeout(5):
                await device.vapix.initialize_users()
                await device.vapix.load_user_groups()
            await device.vapix.initialize_event_instances()

            return device

        except axis.Unauthorized:
            LOGGER.warning(
                "Connected to device at %s but not registered or user not admin.", host
            )

        except (asyncio.TimeoutError, axis.RequestError):
            LOGGER.error("Error connecting to the Axis device at %s", host)

        except axis.AxisException:
            LOGGER.exception("Unknown Axis communication error occurred")

        return device

    def move_go_azi(self, loop, azimuth):
        asyncio.set_event_loop(loop)
        loop.run_until_complete(self.aux_go_azi(self.p, azimuth))

    def move_go_ele(self, loop, elevation):
        asyncio.set_event_loop(loop)
        loop.run_until_complete(self.aux_go_ele(self.p, elevation))

    def get_position(self, loop):
        asyncio.set_event_loop(loop)
        return loop.run_until_complete(self.aux_get_position(self.p))

    async def aux_get_position(self, p):
        return await p.query("position")

    async def aux_go_azi(self, p, azimuth):
        if float(azimuth) > 180.0:
            azi = -360.0 + float(azimuth)
        else:
            azi = float(azimuth)
        await p.control(camera=1, pan=azi)

    async def aux_go_ele(self, p, elevation):
        await p.control(camera=1, tilt=float(elevation))
```

Assuming that the code is correct, can you please instruct me on how to call the classes to reproduce the problem, i.e. show me the main function. There is also the question of the `axis` package. I won't be able to run the code without this package, true?

**Reply by Pedro-Leitek** [reply]
Hi,

def go_azi, def go_ele, def get_pos are outside of the classes. You will need the axis python package from kane610. https://github.com/Kane610/axis.
The main is the following:

`
if __name__=="__main__":
    ai=AIEngine() 
    cmr=ai.CriaCamara(Link, Nome)
    p=ai.CriaPTZ(Link, Nome) 
    t1 = threading.Thread(target=cmr.stream, args=(p,))
    t2 = threading.Thread(target=cmr.obtemFrame, args=(p,))
    t3 = threading.Thread(target=p.getAzimuth)
    t1.daemon = True
    t2.daemon = True
    t3.daemon = True
    t1.start()
    t2.start()
    t3.start()
    server=Server()
    server.run(p)

server is the following:

class Server:
    def __init__(self):
       self.sock=socket.socket(socket.AF_INET, socket.SOCK_STREAM)
       self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
       self.sock.bind((HOST, PORT))
       self.sock.listen()

    def listener(self, p, cmr):
       while True:
          conn, addr = self.sock.accept()     
          while True:
             data = conn.recv(1024)
             if not data:
               break
             data = json.loads(data.decode('utf-8'))
             if data['tipo']=="move":
                  move(p,data['valor'], data['direcao'])
             elif data['tipo']=="mode":
                  cmr.mode=data['valor']
                  if data['valor']=="auto":
                     cmr.tour=data["tour"]
             elif data['tipo']=="sector":
                  cmr.mode="manual"
                  go(p,(int(data["X"])-1)*15,-9*(int(data["Y"])-1))
             elif data['tipo']=="go":
                  go(p,data['valor1'], data['valor2'])
             elif data['tipo']=="zoom":
                  zoom_in_out(p, data['valor'])
             elif data['tipo']=="speeddry":
                  start_speeddry(p)
             elif data['tipo']=="ircutfilter":
                  set_ircutfilter(p, data['valor'])
          
    def run(self,p):
          cThread = threading.Thread(target=self.listener, args=(p,cmr))
          cThread.start()
`

**Reply by mattias-kindborg-at-work** [reply]
I'm falling back to the fact that this would be much easier if you provided the code in a public repository, with instructions on how to install the dependencies, and with clear instructions on how to reproduce the issue. Please let me know when you have this repository ready.

**Reply by Pedro-Leitek** [reply]
Ok, I created a public repository https://github.com/Pedro-Leitek/axis_debug
However this code is dependent of other code. You see, my project is a server-client architecture and the code I'm showing is from client side. It is supposed the server to communicate through network with the client and vice versa. The code I showed is done to work with this architecture. The problem is that I can't provide the full code and put them in a public repository since the code is proprietary from the company I work. But my guess is that in order to reproduce the issue you will only need the function "stream" in auto mode.

**Reply by mattias-kindborg-at-work** [reply]
Thanks for the repository! Can you also add instructions on how to install the dependencies, and what variables I need to change in the source code in order to run this in my network targeting one of my cameras.

**Reply by Pedro-Leitek** [reply]
So, in order to test I would comment anything related to sockets. In fact only the stream thread should be running for testing, as the error is in this thread. You can also comment the "self.processa" function. Make sure the "self.mode" is in "auto". Regarding the packages, you will need asyncio, httpx, opencv.

**Reply by Pedro-Leitek** [reply]
@mattias-kindborg-at-work Hi, do you have any news?

**Reply by mattias-kindborg-at-work** [reply]
Have you updated the repository according to your own comments? Have you added instructions in the repository on how to install all dependencies, and provide me with commands to execute to easily run your code showcasing the problem?

**Reply by Pedro-Leitek** [reply]
@mattias-kindborg-at-work I just updated the repository. Check if you can run the program based on the instructions I put there.

**Reply by mattias-kindborg-at-work** [reply]
Hi @Pedro-Leitek. Did you try to run the code yourself? I'm having no luck since your code still seems to be lacking proper packages and import statements. E.g. you are using the `socket` package but I see no import, you are referencing `cv2` (assuming it to be OpenCV?) but it's not declared in your README as one of your dependencies, nor do I see an import.

Your provided repository is still not something I can use to validate your issue.

**Reply by Pedro-Leitek** [reply]
@mattias-kindborg-at-work I was able to solve the issue. It turns out the repository from kane610 is the problem. If I use the http api commands the issue doesn't happen. Thanks anyway.

**Reply by mattias-kindborg-at-work** [reply]
Great! I'm happy that you persevered and found a fix to your problem!
